### PR TITLE
Re-authenticate if we have the username and password in keychain and get a 401 on launch

### PR DESCRIPTION
**Pull request by orta**
This fixes a long-running problem where the auth token is a 401 from gravity - 
https://github.com/artsy/eigen/issues/2691

Will explain how it works inline

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_travis">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Artsy/App/ARAuthValidator.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * process with stored creds and re-creates the Emission object
+ * with the new access token.
+ */
+@interface ARAuthValidator : NSObject
+
+/// Check for the validity of the current auth token
++ ( void)validateAuthCredentialsAreCorrect;
+
+@end
+
```
Comment (orta) [inline Artsy/App/ARAuthValidator.h]: I added my comments inline instead here ^

File: Artsy/App/ARAuthValidator.m [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        NSString *serverTDLR = [error userInfo][@"NSLocalizedDescription"];
+        if ([serverTDLR containsString:@"401"]){
+            // We hit a 401 so we need to be a bit worried about this
+            ARUserManager *manager = [ARUserManager sharedManager];
+            [manager tryReLoginWithKeychainCredentials:^(User *currentUser) {
+                [[ARAppDelegate sharedInstance] setupEmission];
+            }];
+        }
+    }];
+}
```
Comment (orta) [inline Artsy/App/ARAuthValidator.m]: We don't get a HTTP status code from the NSError, (it's some weird foundation number) so instead grab the error description coming from AFNetworking and check for a 401 in there.

File: Artsy/Constants/ARDefaults.m [hunk 3]
```diff
@@ -9,6 +9,8 @@
 NSString *const ARStagingMetaphysicsURLDefault = @"ARStagingMetaphysicsURLDefault";
 NSString *const ARStagingLiveAuctionSocketURLDefault = @"ARStagingLiveAuctionSocketURLDefault";
 
+NSString *const ARUsernameKeychainKey = @"ARUsernameKeychainKey";
+NSString *const ARPasswordKeychainKey = @"ARPasswordKeychainKey";
 NSString *const AROAuthTokenDefault = @"AROAuthToken";
```
Comment (orta) [inline Artsy/Constants/ARDefaults.m]: We weren't storing username/password at all, except in safari keychain, but that's a special case

File: Artsy/Networking/API_Modules/ARUserManager.m [hunk 4]
```diff
@@ -179,6 +179,7 @@ - (void)loginWithUsername:(NSString *)username
         NSString *expiryDateString = JSON[AROExpiryDateKey];
 
         [ARRouter setAuthToken:token];
+        [self storeUsername:username password:password];
```
Comment (orta) [inline Artsy/Networking/API_Modules/ARUserManager.m]: Makes sure we save on login, this won't work for facebook auth'd folks

File: Artsy/Networking/API_Modules/ARUserManager.m [hunk 5]
```diff
@@ -286,8 +286,8 @@ - (void)loginWithFacebookToken:(NSString *)token
                 networkFailure(error);
             }
         }
+    }];
 
-        }];
     [op start];
 }
```
Comment (orta) [inline Artsy/Networking/API_Modules/ARUserManager.m]: formatting ^

File: Artsy/Networking/API_Modules/ARUserManager.m [hunk 6]
```diff
@@ -442,6 +442,8 @@ + (void)clearUserData:(ARUserManager *)manager useStaging:(id)useStaging
 
     [manager.keychain removeKeychainStringForKey:AROAuthTokenDefault];
     [manager.keychain removeKeychainStringForKey:ARXAppTokenKeychainKey];
+    [manager.keychain removeKeychainStringForKey:ARUsernameKeychainKey];
+    [manager.keychain removeKeychainStringForKey:ARPasswordKeychainKey];
```
Comment (orta) [inline Artsy/Networking/API_Modules/ARUserManager.m]: making sure logout works

File: Artsy/Networking/API_Modules/ARUserManager.m [hunk 7]
```diff
@@ -565,6 +567,33 @@ - (void)saveSharedWebCredentialsWithEmail:(NSString *)email
...
+        // NOOP
+    }
+    gotUser:success
+    authenticationFailure:^(NSError *error) {
+        // NOOP
+    } networkFailure:^(NSError *error) {
+        // NOOP
+    }];
+}
+
```
Comment (orta) [inline Artsy/Networking/API_Modules/ARUserManager.m]: Mostly `NOOP`s but it'll accurately emulate going through the login button process (which will set up the ARRouter etc)

File: Artsy/Networking/API_Modules/ArtsyAPI+CurrentUserFunctions.m [hunk 8]
```diff
@@ -8,6 +8,16 @@
...
 
++ (void)getMeHEADWithSuccess:(void (^)(void))success failure:(void (^)(NSError *error))failure;
+{
+    NSURLRequest *request = [ARRouter newMeHEADRequest];
+    [self performRequest:request success:^(id thing) {
+        success();
+    } failure:^(NSURLRequest *request, NSHTTPURLResponse *response, NSError *error) {
+        failure(error);
+    }];
+}
```
Comment (orta) [inline Artsy/Networking/API_Modules/ArtsyAPI+CurrentUserFunctions.m]: Just exposing a way to get the HEAD for the me route

File: Artsy/App/ARAuthValidator.m (hunk 2 above)
Comment (alloy) [reply Artsy/App/ARAuthValidator.m]: Could it be a wrapped error? (Check `userInfo[NSUnderlyingError]`)

**Comment by alloy** [comment]
Yo, thanks for attempting to deal with this!

The code looks good, although it would be nice to get rid of the string matching of the HTTP status.

However, I’m wondering if it wouldn’t make more sense to have a user re-sign in from the splash screen instead. The reason I say this is because I assume the issue can come up just as well with FB authentication, web credentials, or when a user has changed their password on artsy.net. I think that in all these cases the solution wouldn’t suffice, whereas throwing them back to the splash screen, while slightly less smooth of an experience, would address all of these. Do you have thoughts on this?

**Comment by orta** [comment]
I think I know a reasonable answer, which is to log you out if the re-auth fails or we really can't log you in because of no username/password. 

I've made some changes to do that - it's not elegant because it's an alert and a log out, but the onboarding is meant to be set up at the start and not thrown up async.

**Comment by alloy** [comment]
Merge on green

**Pull request merged**